Switch from builtin video sink to qmlglsink

andrewvoznytsa opened the pull request:
This patch switches QGC from using builtin video sink to qmlglsink (part of GStreamer Good Plug-ins).

This change is part of my work to lower QGC maintenance efforts and to implement standalone test app for video receiver.

Code was tested on Linux (Ubuntu 18.04/Qt 12.6/Gstreamer 1.14.5/qmlglsink - git/master, local build).

Next few days I'll be testing this code on all QGC platforms. So please expect updates (at least for platforms which require static gst linking). Right now it is more like early preview for Linux users - I'd like to hear comments and tweak some parts if there are good reasons.

andrewvoznytsa commented:
In some context - yes.

I'd not do so and save my time but unfortunately - I can't. My problem is that your patch changes some places that are not related to video sink replacement (like noticeable .qml changes), disables/does not implement existing major features (video stream watchdog for example) and keeps object hierarchy still over-complicated (VideoSurface for example - why would we need that with qmlglsink?). I'm trying to simplify VideoReceiver and keep all existing features working as before.

Right now I'm trying to make it working on all QGC platforms and looking to start working on standalone test tool for VideoReceiver - that's annoying to see many video issues and have no instrument for quick automated testing.

As for me it looks like that my patch is much closer to release candidate than yours. I propose to use it as basis for qmlglsink integration. If you see any problems with my code - let me know and I'll fix ASAP.

tcanabrava commented:
I don't mind working in a feature together with others, please don't think I'm complaining. My idea with my patch series is not just to remove the QtGstreamer code but also so simplify the whole architecture. You can see for instance that I managed to get multiple video streams working with it while the original architecture for video in QGC makes this quite difficult. 

The VideoSurface is an abomination that I created to be able to start / stop the video streams in the Qml rendering thread, as I was not having videos with a pipeline started outside of it. It's probably something uneeded and that I could rewrite in a better way, I just need to found out how.

On your side, you used the example code for the qmlsink that uses "objectName" on the Qml - that's one thing that I did not want to do on my code as it's quite unusual for Qml objects to have a name, and feels wrong.

andrewvoznytsa commented:
Yes, I see your multiple video streams feature and I think that it is something that maintainers would adopt. From your past communications with them it sounds like they prefer smooth transition from existing code to full featured multiple view support (i.e. small step increments keeping QGC not worse than it was before). I think I'm finishing one of these small steps (qmlglsink support) and probably it is where multiple video stream support can start from.

Regarding object names - I agree. Adding extra field is not very nice solution. But since it requires two extra lines only I think that it is not so bad for current stage. If you see some nice way to workaround this requirement then drop me a message (you can email me directly) and I'll be happy to add it to my patch.

tcanabrava commented:
I just did a revamp of the "qmlglsink" also changing a few bits, I would really appreciate a review from somebody that knows more of GStreamer than I do. Even if we ship your version instead of mine first - as mine deals with more changes in the code architecture.

I *believe* I can transform the VideoSurface code into the VideoReceiver (the only difference is that I would need to remove the VideoReceiver from the VideoManager, and initialize it directly in Qml for the threading issue I commented about) - would will also make things better: VideoManager is bloated and tries to manage things outside of it's scope, so getting rid of it seems a plus. 

https://github.com/mavlink/qgroundcontrol/pull/8071

About the "objectName" - it does not only changes two lines in code, you also need to get the rootWindow() and walk thru the object hierarchy to find the ones you are looking for (mostly making impossible to transform the video code in a single app for testing purposes) - that can be problematic if two QObjects have the same name.

andrewvoznytsa commented:
I'll look on objectName after finishing with non-Linux supports - qmlglsink is not as smooth as it looks on first sight.

andrewvoznytsa commented:
Latest changes to this PR are adding prebuilt and statically linked qmlglsink to the QGC. It makes QGC working* on Linux, Windows**, macOS, iOS and Android. NOTE that QGC tries to use qmlglsink installed on your system. If qmlglsink is not present - app will fallback to statically linked qmlglsink and will use it.

*working - I mean good enough for development use. QGC still misses a few small fixes here and there. One of them - after initial app setup and first time video streaming configuring you need to restart app to make streaming working. It should be trivial to fix and I'll do that shortly. iOS build sometimes crashes during startup. I'm not sure that it is related to qmlglsink. But I'm also not sure that it is not related to that.

**Windows - qmlglsink depends on Open GL which was never simple to use on Windows. There is some long lasting bug related to multiple video sink support (see https://gitlab.freedesktop.org/gstreamer/gst-plugins-good/issues/477#note_361958).  You need to disable #if at src/libs/gst-plugins-good/ext/qt/gstqtglutility.cc@224 if you experience issues with video on Windows. I'll continue working on this issue and fix it.

andrewvoznytsa commented:
Fixed video initialization on first startup. I'd ask someone with deep QGC/QML understanding to look on that fix (see https://github.com/andrewvoznytsa/qgroundcontrol/commit/886c9686d96fbfcdfbc3de003eb34f0345d7e06f) - basically videoBackgroundComponent has the same life cycle as thermalItem and there should be no any problems. I had to do that to keep qmlglsink initialization simple, in single place on startup.

Fixed qmlglsink multiple instances support on Windows. PR sent to upstream (see https://gitlab.freedesktop.org/gstreamer/gst-plugins-good/merge_requests/432) and is under review. I'll continue working on that until it will be fully accepted.

From now I expect that this patch makes QGC not worse (feature wise) than it was before qmlglsink switch and it is in good shape to be considered for merge.

Future plans - I plant to remove objectName dependecy as @tcanabrava suggested. It will make video receiver easier to use in different use cases, including multiple video stream support. Also I'll try to improve videoBackgroundComponent initialization and make it on demand, to better manage resources, probably improve startup time and make things right. But it all depends on my other work related to VideoReceiver improvements. I expect to refactor VideoReceiver a bit (and probably VideoManager). And I think that these objectName and intializations improvements will land into repo as separate PRs, during future work.

tcanabrava commented:
Andrew, I'm trying to follow the commits but it's quite hard with the way you use git, can you perhaps clear the commits and use git rebase? The current state of the commits is not really easy to follow.

andrewvoznytsa commented:
One of private 'quick' commits escaped to public repo some time ago and rebasing is not trivial at this point :( Will see what can I do after fixing CI issues.

jaxxzer commented:
> I expect to refactor VideoReceiver a bit (and probably VideoManager).

Hi @andrewvoznytsa. Please consider coordinating with our efforts to refactor the video code by elaborating your plans and ideas at https://github.com/mavlink/qgroundcontrol/issues/7842.

andrewvoznytsa commented:
Hi @jaxxzer - sure, before starting to work on next massive update I'll post details at https://github.com/mavlink/qgroundcontrol/issues/7842. Basically I'm working according to your list, +-.

DonLakeFlyer commented:
@andrewvoznytsa FYI: @rollysing Has a 100% crash case on current daily gstreamer. So you can also use home to test that it has been fixed with you changes.

andrewvoznytsa commented:
@DonLakeFlyer, @rollysing I'd fix that crash of course but need more info. Does this PR branch https://github.com/mavlink/qgroundcontrol/pull/8093 also crashes? How can I reproduce that crash on my side?

DonLakeFlyer commented:
I don't think @rollysing has the ability to build QGC. You'll need to provide him with a Windows test installer to run.

The background is that there are a set of Lenovo laptops which use the Intel integrated graphics OpenGL driver. On this subset of machines all you need to do is boot QGC with video enabled (any kind) and it will crash on boot. @rollysing seems to have one of these laptops. And @antiheavy has a number of Sentera customers with these. All of them crash on boot.

andrewvoznytsa commented:
Sorry for long silence - New Year is always hard period for me.

Let's continue with Lenovo - there https://drive.google.com/file/d/1AuRpDDivWl1XraD90SwUY--Bsx7DvXPg/view?usp=sharing is custom QGC build with qmlglsink support.

@rollysing and @Antiheavy - if you could - please test this build on your Lenovos and let me know if you still see that mystic crash.

rollysing commented:
@andrewvoznytsa No joy. Watching in Task Manager, it starts then Status will display Suspended then the process disappears.

Antiheavy commented:
> Let's continue with Lenovo - there https://drive.google.com/file/d/1AuRpDDivWl1XraD90SwUY--Bsx7DvXPg/view?usp=sharing is custom QGC build with qmlglsink support.
> 
> @rollysing and @Antiheavy - if you could - please test this build on your Lenovos and let me know if you still see that mystic crash.

This test build you linked installs and loads fine with the default settings (video disabled), but when I try to enable TCP video stream the application locks up ("not responding") and must be force closed with windows task manager.  Today's QGC Daily Build seems to work fine and does not lock up when I enable a TCP video stream.

rollysing commented:
No joy on any startup mode of the Daily Build either.

andrewvoznytsa commented:
If you could - please use UDP source. TCP is broken in this branch. Or, if
you prefer, I’ll merge fix and submit new build on Jan 8.

On Mon, 6 Jan 2020 at 17:34, Todd Colten <notifications@github.com> wrote:

> Let's continue with Lenovo - there
> https://drive.google.com/file/d/1AuRpDDivWl1XraD90SwUY--Bsx7DvXPg/view?usp=sharing
> is custom QGC build with qmlglsink support.
>
> @rollysing <https://github.com/rollysing> and @Antiheavy
> <https://github.com/Antiheavy> - if you could - please test this build on
> your Lenovos and let me know if you still see that mystic crash.
>
> This test build you linked installs and loads fine with the default
> settings (video disabled), but when I try to enable TCP video stream the
> application locks up ("not responding") and must be force closed with
> windows task manager. Today's QGC Daily Build seems to work fine and does
> not lock up when I enable a TCP video stream.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mavlink/qgroundcontrol/pull/8093?email_source=notifications&email_token=ADKJSBLACLAZCQACIOSQAZLQ4NFR3A5CNFSM4JV65OG2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEIFZR4I#issuecomment-571185393>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADKJSBLS4MS73HWSBNIBGQTQ4NFR3ANCNFSM4JV65OGQ>
> .
>

andrewvoznytsa commented:
OK - I’ll try to add some kind of crash dump to see where and why it is
crashing.

On Mon, 6 Jan 2020 at 18:18, rollysing <notifications@github.com> wrote:

> No joy on any startup mode of the Daily Build either.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mavlink/qgroundcontrol/pull/8093?email_source=notifications&email_token=ADKJSBK5OQVJ33OIZKNJXHTQ4NKUXA5CNFSM4JV65OG2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEIF5Z3Y#issuecomment-571202799>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADKJSBJTWR34MIYHDEAVOSTQ4NKUXANCNFSM4JV65OGQ>
> .
>

rollysing commented:
@andrewvoznytsa It doesn't even get to that point. QGC crashes during opening. It only stays in Task Manager for a few seconds then disappears.

andrewvoznytsa commented:
Sounds like you have mix if 32 bit and 64 bit dlls but it is just blind
guess... I’ll try to find some solution.

On Mon, 6 Jan 2020 at 23:20, rollysing <notifications@github.com> wrote:

> @andrewvoznytsa <https://github.com/andrewvoznytsa> It doesn't even get
> to that point. QGC crashes during opening. It only stays in Task Manager
> for a few seconds then disappears.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mavlink/qgroundcontrol/pull/8093?email_source=notifications&email_token=ADKJSBOMBNL6274X55MS6C3Q4OODHA5CNFSM4JV65OG2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEIG2KKA#issuecomment-571319592>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADKJSBLO763AS3SSEDMBK33Q4OODHANCNFSM4JV65OGQ>
> .
>

andrewvoznytsa commented:
I believe macOS issue reported by Travis (see https://travis-ci.org/mavlink/qgroundcontrol/builds/626084593?utm_source=github_status&utm_medium=notification) is due to outdated gstreamer package. I'll try to fix Travis...

andrewvoznytsa commented:
@Antiheavy there https://drive.google.com/file/d/1iPpPsqq_pJP7bpaDiYsLQv-e0OCcitF-/view?usp=sharing you can get new build with TCP support (merged fix from master).  Please try on your side and let me know if you see any issues.

rollysing commented:
This is still no joy for me, all three iterations of the app.

Thanks.

andrewvoznytsa commented:
@rollysing this build has no improvements for your case - just fixed TCP issues for @Antiheavy to allow him to test that code. As for you then I suppose that you have some system configuration issues. What you say sounds like QGC pickups 32 bit and 64 bit dlls and crashes since it is not valid way to load for 32 bit app. To stop guessing and have real information I need to implement some crash dump reporting to see what is happening. I see that there is open issue https://github.com/mavlink/qgroundcontrol/issues/4445 related to crash data collection. We need to wait until this issue will be resolved or, if teamviewer works for you, then we can agree some debugging session on your side. If teamviewer works for you then you can email me privately and we'll agree details - this is probably fastest way to solve that issue.

Antiheavy commented:
> @Antiheavy there https://drive.google.com/file/d/1iPpPsqq_pJP7bpaDiYsLQv-e0OCcitF-/view?usp=sharing you can get new build with TCP support (merged fix from master). Please try on your side and let me know if you see any issues.

@andrewvoznytsa I installed the version linked above and successfully tested on our hardware in the lab (Lenovo windows laptop with TCP video connection to our drone).  Seems to behave the same as upstream QGC, which I think was the goal(?).

![image](https://user-images.githubusercontent.com/12117027/72743808-fa020100-3b71-11ea-9b51-4a98d7e37444.png)

andrewvoznytsa commented:
> > @Antiheavy there https://drive.google.com/file/d/1iPpPsqq_pJP7bpaDiYsLQv-e0OCcitF-/view?usp=sharing you can get new build with TCP support (merged fix from master). Please try on your side and let me know if you see any issues.
> 
> @andrewvoznytsa I installed the version linked above and successfully tested on our hardware in the lab (Lenovo windows laptop with TCP video connection to our drone). Seems to behave the same as upstream QGC, which I think was the goal(?).
> 
> ![image](https://user-images.githubusercontent.com/12117027/72743808-fa020100-3b71-11ea-9b51-4a98d7e37444.png)

@Antiheavy please read below:

From @DonLakeFlyer 

> I don't think @rollysing has the ability to build QGC. You'll need to provide him with a Windows test installer to run.
> 
> The background is that there are a set of Lenovo laptops which use the Intel integrated graphics OpenGL driver. On this subset of machines all you need to do is boot QGC with video enabled (any kind) and it will crash on boot. @rollysing seems to have one of these laptops. And @Antiheavy has a number of Sentera customers with these. All of them crash on boot.

As far as I understood you and @rollysing have the same Lenovo laptops that were always crashing. I sent you my builds to verify if these crashes are still there. Sound like last one is working on your side and there are no issue. Strange thing for me is that you say that upstream QGC is working for you too. Then I'm not sure what was not working - miscommunication?

Antiheavy commented:
> miscommunication?

maybe slightly, but it is true that we've had a large number of users with windows laptops (typically Lenovo's) where QGC crashes seemingly randomly during use.  @DonLakeFlyer narrowed it down to a seeming issue with video drivers (typically intel integrated GPU) and Gstreamer.  This issue seems different than @rollysing because his sounds like a repeatable crash on boot (?).  Our issue is/was random crashes during use (sometimes after 20 minutes, sometimes after hours).  Some of our customer's laptops seemed to crash daily, however, our internal laptops that we have access too would only crash occasionally.  I think the only way for us to test if our random crashing issue is improved with the newer QGC is through long term use over time.

andrewvoznytsa commented:
Got it. And just to confirm - you don’t see gstreamer-related crashes in my
build for now.

On Thu, 23 Jan 2020 at 16:40, Todd Colten <notifications@github.com> wrote:

> miscommunication?
>
> maybe slightly, but it is true that we've had a large number of users with
> windows laptops (typically Lenovo's) where QGC crashes seemingly randomly
> during use. @DonLakeFlyer <https://github.com/DonLakeFlyer> narrowed it
> down to a seeming issue with video drivers (typically intel integrated GPU)
> and Gstreamer. This issue seems different than @rollysing
> <https://github.com/rollysing> because his sounds like a repeatable crash
> on boot (?). Our issue is/was random crashes during use (sometimes after 20
> minutes, sometimes after hours). Some of our customer's laptops seemed to
> crash daily, however, our internal laptops that we have access too would
> only crash occasionally. I think the only way for us to test if our random
> crashing issue is improved with the newer QGC is through long term use over
> time.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mavlink/qgroundcontrol/pull/8093?email_source=notifications&email_token=ADKJSBKRZRZ2NGOBF4UCBVLQ7GT5TA5CNFSM4JV65OG2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJXSOWY#issuecomment-577709915>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADKJSBKAOSJIW6EQ4ASPSB3Q7GT5TANCNFSM4JV65OGQ>
> .
>

rollysing commented:
@Antiheavy @andrewvoznytsa Just in case you have been aware, @DonLakeFlyer made me test QGroundControl Development NoGSTNoUVC:33931657b 2019-07-04 10:39:58 -0700 and thus far is the only version that works for me. Hope this helps.

Thanks.

andrewvoznytsa commented:
@rollysing would it be possible to install WinDBG https://docs.microsoft.com/ru-ru/windows-hardware/drivers/debugger/debugger-download-tools, run QGC and do a few simple steps to get us back trace (I'll describe steps if WinDBG is something that you can do)?

Antiheavy commented:
> Got it. And just to confirm - you don’t see gstreamer-related crashes in my build for now.

I did not, but I just did a basic test in the lab.  I did not "endurance" test it.

rollysing commented:
> @rollysing would it be possible to install WinDBG https://docs.microsoft.com/ru-ru/windows-hardware/drivers/debugger/debugger-download-tools, run QGC and do a few simple steps to get us back trace (I'll describe steps if WinDBG is something that you can do)?

I've downloaded the app. Please provide me step-by-step on what to do and will get it done as soon as I could.

DonLakeFlyer commented:
FYI: Windows builds installed using the standard installer already set windows up to capture crash mini dumps. The docs here: https://docs.qgroundcontrol.com/en/Support/Support.html explain where the dump files are stored. You can then load these in visual studio debugger and get a stack track of where the crash is.

andrewvoznytsa commented:
@rollysing could you please do what @DonLakeFlyer suggest in https://github.com/mavlink/qgroundcontrol/pull/8093#issuecomment-583161395 ? I was not aware of that. Please send me these dump files (*.dmp) and I'll check them here. If we don't get enough information then we'll continue with WinDBG. To find dumps - please look for 'Reporting Crashes from Windows Builds' on https://docs.qgroundcontrol.com/en/Support/Support.html page.

DonLakeFlyer commented:
> If we don't get enough information then we'll continue with WinDBG.

With a dump file you'll get way more than what you could get remotely getting rolly to follow steps. If a mini-dump doesn't get you enough information. You can ask rolly to futz the registry to get a full dump. A full dump will get you the same thing as if you had captured the crash in the debugger on your own machine. Move around the stack, examine variable and so forth.

DonLakeFlyer commented:
The Windows installer write the following reg keys:
```
  WriteRegDWORD HKLM "SOFTWARE\Microsoft\Windows\Windows Error Reporting\LocalDumps\${EXENAME}.exe" "DumpCount" 5 
  WriteRegDWORD HKLM "SOFTWARE\Microsoft\Windows\Windows Error Reporting\LocalDumps\${EXENAME}.exe" "DumpType" 1
  WriteRegExpandStr HKLM "SOFTWARE\Microsoft\Windows\Windows Error Reporting\LocalDumps\${EXENAME}.exe" "DumpFolder" "%LOCALAPPDATA%\QGCCrashDumps"
```

If you change DumpType in the registry to 2 you get a full dump. Start with a mini which is default though since full dumps are huge.

rollysing commented:
@andrewvoznytsa Sorry. I've been out of town. Here are the crash dumps of a 02/20/2020 Daily Build. It opened up fine after initial installation. The moment I changed the video option to H.264, it crashed.
https://1drv.ms/u/s!Ar66Lk6yrxscgyTSNd6KuslMDDNT?e=Ct4pmx

Hope this helps. Let me know if there's anything you want me to do.

Thanks.

andrewvoznytsa commented:
Thanks for update! That moment, related to video code change, was probably fixed and merged into master. If you get Daily build dated later than this comment then you can expect that there is no crash anymore (I hope very much). Could you please check? In the mean time I'll look on crash dump.

rollysing commented:
@andrewvoznytsa The crash dumps I referenced are from a Daily Build I just downloaded and installed 2 hours ago. So the merge is still crashing once I opted to enable the video option.

andrewvoznytsa commented:
![image](https://user-images.githubusercontent.com/13932805/74248451-431c1f80-4cf0-11ea-8cab-0e5fb5ae8b11.png)

It crashes somewhere inside g_slice_alloc() (malloc() equivalent, part of glib which is part of gstreamer). Everything was built without debug info - that's piece of work to rebuild all things with debug info to get something more useful than what I've found (manual stack walking is possible of course, but IMO it has no sense in our case).

If we want to save time then local debugging via Teamviewer can help.

rollysing commented:
@andrewvoznytsa I'm UTC -05:00 US Eastern Time. Let me know when you'd be available and I'll have it setup and available for you to remote in.

Thanks.

andrewvoznytsa commented:
@rollysing I'll prepare for this test and let you know (if we do this immediately I'll need to setup all dev tools which I usually use - this does not sound for me as right path to ask you install all these things - will need to see how can I debug that with minimal debug tool set and quickly). 

In the mean time, we can continue our lottery. I just pushed https://github.com/mavlink/qgroundcontrol/pull/8318 which sounds close to what you were reporting. I've doubts that it is also your issue but we can try - it sounds close. I expect that it will be merged shortly and you can try that tomorrow.

DonLakeFlyer commented:
A full dump instead of a mini-dump won't provide you enough information? That is about the same as getting it in the debugger yourself.

andrewvoznytsa commented:
The problem here is that gstreamer and glib were built without debug info.
VC is unable to show backtrace. I think that full dump also wont help.

Need to try on my side to see what is the easiest way to find what we are
looking for.

On Wed, 12 Feb 2020 at 19:41, Don Gagne <notifications@github.com> wrote:

> A full dump instead of a mini-dump won't provide you enough information?
> That is about the same as getting it in the debugger yourself.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mavlink/qgroundcontrol/pull/8093?email_source=notifications&email_token=ADKJSBOPW2AXSHWV44R7I2LRCQYDNA5CNFSM4JV65OG2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOELRWJMY#issuecomment-585327795>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADKJSBOHOREDJLJGMBNAJ3DRCQYDNANCNFSM4JV65OGQ>
> .
>

andrewvoznytsa commented:
@rollysing here is debug build for you - https://drive.google.com/open?id=1eGWe-X6fv7n5fW_elpA13on-GqUEClfM

It is latest master built in debug mode with symbols.

You'll need to unpack that to C:\. After all you'll have C:\lenovo folder with build-qgroundcontrol-Desktop_Qt_5_12_6_MSVC2017_64bit-Debug and qgroundcontrol subfolders. This is large stuff - uses about 2 Gb.

As first step we can try naive approach - let's run c:\lenovo\build-qgroundcontrol-Desktop_Qt_5_12_6_MSVC2017_64bit-Debug\debug\QGroundControl.exe and try to catch dump. Then send it to me - hopefully I'll be able to find all info there.

If no - we'll need to continue with teamviewer.

I'm here for next 4-6 hours - should be enough to fix that. If you are busy today - we can continue any other day.

rollysing commented:
Getting this error almost immediately after starting installation:
![image](https://user-images.githubusercontent.com/20650795/75376575-468dda00-589e-11ea-9fe6-9fd91f7cf469.png)

dogmaphobic commented:
Debug builds only run on a system with all the debug DLLs loaded.

andrewvoznytsa commented:
Sigh. I'll prepare better build - will setup clean VM, collect all required DLLs and post it here.

andrewvoznytsa commented:
@rollysing please find missing dll there https://drive.google.com/open?id=1AQYHg3CX7rIj7V-RJpNmkHPM-4riyHUG
You need to unpack it into `c:\lenovo\build-qgroundcontrol-Desktop_Qt_5_12_6_MSVC2017_64bit-Debug\debug`, next to QGroundControl.exe
Then it should work - at least worked for me on fresh Windows system.

Next step would be to run WinDBG for AMD64
![Windows 10 x64-2020-02-27-20-59-52](https://user-images.githubusercontent.com/13932805/75477945-b935ad00-59a5-11ea-96ea-1e49f8186b59.png)
Then open open executable QGroundControl.exe like there
![Windows 10 x64-2020-02-27-21-00-24](https://user-images.githubusercontent.com/13932805/75477999-d8ccd580-59a5-11ea-8bc4-aa01afb54aaa.png)
After that you'll see something like to this screen
![Windows 10 x64-2020-02-27-21-00-44](https://user-images.githubusercontent.com/13932805/75478067-f8fc9480-59a5-11ea-9649-8647a8abc969.png)
This is time to run app pressing F5 and get to next point
![Windows 10 x64-2020-02-27-21-01-29](https://user-images.githubusercontent.com/13932805/75478151-1df10780-59a6-11ea-9018-aa20840854fd.png)
That screen tells us that app crashed so we need to look at call stack (Alt+6)
![Windows 10 x64-2020-02-27-21-01-52](https://user-images.githubusercontent.com/13932805/75478263-5264c380-59a6-11ea-836d-fdde613abafb.png)
Finally we see call stack and can examine what was wrong. So I'm double clicking on the most reasonable line (first QGroundControl.exe enetry) and looking for code which caused that crash
![Windows 10 x64-2020-02-27-21-02-07](https://user-images.githubusercontent.com/13932805/75478362-85a75280-59a6-11ea-85b7-85806848625d.png)
and it tells me obvious thing - someone like me modified _pipeline pointer to point to the middle of nowhere to simulate crash, to be able to write that short guide and verify everything.

Now it is your turn :)

If this looks like alien technology - give me please teamviewer access and I'll debug it myself (perhaps we can use @DonLakeFlyer as man in the middle to send logins details or other contact information).

DonLakeFlyer commented:
Don't see why crash dump file isn't the simplest way to go where. With a full crash dump for example you get everything you could do with WinDbg. See the stack trace at crash, move around the stack, look at variables/memory at each point in the stack, link to the source.

andrewvoznytsa commented:
@rollysing you can simply send crash dump after unpacking that missing dll if that is easier for you - did not wanted to throw all that WinDbg stuff on your head - it just looked safer for me to get needed crash information via WinDbg and finally fix that issue. We can always return to WinDbg if I'll be unable to find required debug info from crash dump because of some unforeseen problem (most likely everything will be good this time).

DonLakeFlyer commented:
By default the QGC Windows installer will set up the registry to collect mini dumps. That is the way to start. If from that you don't get enough information you can switch to full dumps (they are huge).

I uploaded some reg files to S3 which will switch between the QGC default and full dumps:
* https://qgroundcontrol.s3-us-west-2.amazonaws.com/CrashDumpDefault.reg
* https://qgroundcontrol.s3-us-west-2.amazonaws.com/CrashDumpFull.reg

rollysing commented:
@andrewvoznytsa @DonLakeFlyer Good news! That QGC didn't crash! More importantly, I'm able to open a video overlay and the USB VTx is showing it! There's no crash dump to send so is there a log in this that you want? So, the next step is to incorporate what you have here into the stable version?Let me know.

Thanks.

andrewvoznytsa commented:
@rollysing don't know if I should be happy or worried even more - let's try latest 4.0 daily build on your side and see if video is still working for you.

There are possible two kind of problems:
1) there is some difference in debug and release builds (what works for you was debug build) - it will be hard to fix (actually just time consuming and this is what worries me)
2) one of latest changes fixed that issue on your side - in this case we'll live with current code until you see this bug again. If we reproduce it once more we'll have to do binary search between commits that is known to work and this which is crashing to find change that fixes it.

rollysing commented:
@andrewvoznytsa Good news! I just downloaded the daily build and it installed with no problem, opens and am able to have the USB 2.0 Camera show on the overlay.

andrewvoznytsa commented:
@rollysing Perfect! Please let me know ASAP as I'll break that :)

BTW, did you update something on your system recently? Perhaps older builds work too?

rollysing commented:
@andrewvoznytsa Good news again! Installed stable 4.0 and it's working! So it looks like she's good from this stable version and on. Next step would be a field test to see how she does.

Any updates would've been the usual Windows patch Tuesday stuff. So if there's anything from the past ones that you think might have contributed to this success would be it.

In the meantime, will report back (don't know when) on the field test. Thank you very much for all the patience and assistance.

andrewvoznytsa commented:
@rollysing Good luck with your field test!
